ADR/00xx-related-pipelines.md

```diff
@@ -0,0 +1,104 @@
+
+### Pipelines as Code (PaC)
+
+PaC will expand its design to include orchestration logic. This orchestrator will enable a plugable scheduler which is capable of identifying the dependent repositories and Tekton definitions (composing the metadata required to specify a component in AppStudio). When a component's Tekton definition indicates a dependency on another component, PaC will be able to combine the defined pipelines so that the resulting Pipeline definition builds images in the proper order.
+
+When identifying the pipeline definitions to merge, PaC will first look at the linked components for the same branch where a build was triggered from. If no matching branches are found then the default branch will be used.
+
+PaC will need to recursively expand the DAG to ensure that all dependent components are linked.
+
+### Build service
+
+The build service will be able to detect any component pull specs from earlier in a composite pipeline that differ from a previous Snapshot and replace all references of the previous pull spec with the new pull spec before building the component.
+
+### Integration service
+
+The integration service will be able to support updating multiple Component updates from a single PipelineRun such that all components built in a composite pipeline can be tested together as an Application in any configured integration tests.
+
+### Eventual consistency
+
+This flow of related changes as proposed in this ADR is limited to changes in the pull request Tekton definitions. It enables related components to be rebuilt as needed to ensure that image references are maintained and up to date. Once pull requests are merged for one component and the resulting container image is built, the resulting Snapshot will be "out of sync" in that some references in dependent images (i.e. RelatedImages in bundles or image references in a Dockerfile) will not match the latest artifact in the Snapshot. When build dependencies within the same Snapshot are enabled ([RHTAP-967](https://issues.redhat.com/browse/RHTAP-967)), this synchronization issue will be detected and relevant policies will fail.
+
+After a new Snapshot is produced, a Renovatebot will be responsible to propogating the updated image references to all other Components within the Application. After these pull requests have all been merged, consistency should be achieved again and all relevant policies should pass. 
+
+## Applied to Use Cases
+
+Let's apply the architecture to some use cases, and see how it plays out:
+
+### Common Parent Image
+
+Scenario: an application image depends on a common parent image. The user has 1 Application, and 2 Components. One of them is the parent image. The other image is built FROM that image.
+
+* The common parent image will register a dependency to all other Components
```

Comment:
I have an issue with the usage of the word "dependency". If I say "A depends on B", B can be built without A, but A needs B to be built. I.e if I have no dependencies, I can be built first.

This document talks about building a graph, and suggest that the root node(s) depend on the children nodes. This went against my expectation, and I found it confusing.

Maybe we can use graph terms? E.g. `Components directly impacted by a rebuild of the parent will be registered as children in the parent definition`?

Reply:
I think I might understand your issue with the word "dependency" here. I was thinking that A depends on B to test, but confusion can arise when those words are missing. I think clarity could be improved by illustration with a graph, but I also feel like the following representation is too much:

> Each component will be able to describe the builds which need to be triggered after its completion in order to properly test. When a PR build is started for a specific component, the triggered builds will be added as children to the current node. A DAG will be created by visiting each node and registering the children nodes based on the triggering relationships encoded. Once recursion has ceased, execution can proceed by building all Components in the next layer whose every parent has previously been built.

---

> This document talks about building a graph, and suggest that the root node(s) depend on the children nodes. This went against my expectation, and I found it confusing.

What is the expectation you are referring to? The use of the dependency concept again or related to the graph construction?

Reply:
How would the child image be able to display its dependency? If I'm looking at the child image in the UI, how does HAC know to represent the dependency noted on the parent image? 

I had a similar comment to Romain's here, but on Ralph's proposal in #113 for 'depends on', in that it seemed intuitively backwards to me. From a graph perspective, it makes more sense to me to think of the child image depending on the parent, since it needs it to build, although I understand that we want a change to the parent to test the child, and the reverse wouldn't be true.